File: implementations/rust/examples/error_handling.rs
use jyaml::parse;

fn main() {
    println!("=== JYAML Error Handling Examples ===\n");

    // Example 1: Syntax error
    println!("1. Syntax error (unclosed string):");
    let bad1 = r#""unclosed string"#;
    match parse(bad1) {
        Ok(value) => println!("  Unexpected success: {:?}", value),
        Err(e) => println!("  Error: {}\n", e),
    }

    // Example 2: Invalid number format
    println!("2. Invalid number format (leading zero):");
    let bad2 = "01234";
    match parse(bad2) {
        Ok(value) => println!("  Unexpected success: {:?}", value),
        Err(e) => println!("  Error: {}\n", e),
    }

    // Example 3: Tab in indentation
    println!("3. Tab character in indentation:");
    let bad3 = "\tindented with tab";
    match parse(bad3) {
        Ok(value) => println!("  Unexpected success: {:?}", value),
        Err(e) => println!("  Error: {}\n", e),
    }

    // Example 4: Trailing comma in array (now valid in JYAML 0.3)
    println!("4. Trailing comma in array:");
    let good4 = "[1, 2, 3,]";
    match parse(good4) {
        Ok(value) => println!("  Success: {:?}\n", value),
        Err(e) => println!("  Unexpected error: {}\n", e),
    }

    // Example 5: Trailing comma in object (now valid in JYAML 0.3)
    println!("5. Trailing comma in object:");
    let good5 = r#"{"name": "test", "age": 30,}"#;
    match parse(good5) {
        Ok(value) => println!("  Success: {:?}\n", value),
        Err(e) => println!("  Unexpected error: {}\n", e),
    }

    // Example 6: Duplicate keys
    println!("6. Duplicate keys in object:");
    let bad6 = r#"
"name": "first"
"age": 25
"name": "second"
"#;
    match parse(bad6) {
        Ok(value) => println!("  Unexpected success: {:?}", value),
        Err(e) => println!("  Error: {}\n", e),
    }

    // Example 7: Invalid escape sequence
    println!("7. Invalid escape sequence:");
    let bad7 = r#""invalid \x escape""#;
    match parse(bad7) {
        Ok(value) => println!("  Unexpected success: {:?}", value),
        Err(e) => println!("  Error: {}\n", e),
    }

    // Example 8: Invalid Unicode escape
    println!("8. Invalid Unicode escape:");
    let bad8 = r#""invalid \uGGGG""#;
    match parse(bad8) {
        Ok(value) => println!("  Unexpected success: {:?}", value),
        Err(e) => println!("  Error: {}\n", e),
    }

    // Example 9: BOM at beginning
    println!("9. BOM (Byte Order Mark) at beginning:");
    let bad9 = "\u{FEFF}test";
    match parse(bad9) {
        Ok(value) => println!("  Unexpected success: {:?}", value),
        Err(e) => println!("  Error: {}\n", e),
    }

    // Example 10: Invalid identifier
    println!("10. Invalid identifier:");
    let bad10 = "yes"; // YAML allows 'yes' as true, but JYAML doesn't
    match parse(bad10) {
        Ok(value) => println!("  Unexpected success: {:?}", value),
        Err(e) => println!("  Error: {}\n", e),
    }

    // Example 11: Missing colon in object
    println!("11. Missing colon in object:");
    let bad11 = r#"{"name" "value"}"#;
    match parse(bad11) {
        Ok(value) => println!("  Unexpected success: {:?}", value),
        Err(e) => println!("  Error: {}\n", e),
    }

    // Example 12: Unescaped control character
    println!("12. Unescaped control character:");
    let bad12 = "\"line1\nline2\""; // Unescaped newline
    match parse(bad12) {
        Ok(value) => println!("  Unexpected success: {:?}", value),
        Err(e) => println!("  Error: {}\n", e),
    }

    println!("=== Error handling examples completed! ===");
}

File: implementations/python/examples/error_handling.py
#!/usr/bin/env python3
"""JYAML error handling examples."""

from jyaml import loads, parse
from jyaml.lexer import LexerError
from jyaml.parser import ParseError

def demonstrate_lexer_errors():
    """Show various lexer errors."""
    print("=== Lexer Errors ===")
    
    # Tab character error
    try:
        loads('{\t"key": "value"}')
    except (LexerError, ParseError) as e:
        print(f"Tab error: {e}")
    
    # Unterminated string
    try:
        loads('{"unterminated": "string')
    except (LexerError, ParseError) as e:
        print(f"Unterminated string: {e}")
    
    # Invalid escape sequence
    try:
        loads('{"invalid": "\\q"}')
    except (LexerError, ParseError) as e:
        print(f"Invalid escape: {e}")
    
    # Unknown identifier
    try:
        loads('{"value": undefined}')
    except (LexerError, ParseError) as e:
        print(f"Unknown identifier: {e}")
    
    print()

def demonstrate_parser_errors():
    """Show various parser errors."""
    print("=== Parser Errors ===")
    
    # Missing value
    try:
        loads('{"key": }')
    except ParseError as e:
        print(f"Missing value: {e}")
    
    # Mismatched brackets
    try:
        loads('[1, 2, 3}')
    except ParseError as e:
        print(f"Mismatched brackets: {e}")
    
    # Missing comma
    try:
        loads('{"a": 1 "b": 2}')
    except ParseError as e:
        print(f"Missing comma: {e}")
    
    print()

def demonstrate_graceful_handling():
    """Show how to handle errors gracefully."""
    print("=== Graceful Error Handling ===")
    
    test_inputs = [
        '{"valid": "json"}',
        '{"invalid": }',
        'null',
        '[1, 2, 3,]',  # Valid with trailing comma
        '{\t"tab": "error"}',
    ]
    
    for i, input_str in enumerate(test_inputs, 1):
        print(f"Test {i}: {input_str[:20]}...")
        try:
            result = loads(input_str)
            print(f"  ✓ Success: {result}")
        except (LexerError, ParseError) as e:
            print(f"  ✗ Error: {type(e).__name__}: {e}")
        except Exception as e:
            print(f"  ✗ Unexpected error: {type(e).__name__}: {e}")
        print()

def validate_input(input_str: str) -> bool:
    """Validate JYAML input without raising exceptions."""
    try:
        loads(input_str)
        return True
    except (LexerError, ParseError):
        return False

def main():
    demonstrate_lexer_errors()
    demonstrate_parser_errors()
    demonstrate_graceful_handling()
    
    # Validation example
    print("=== Input Validation ===")
    test_cases = [
        '{"valid": true}',
        '{"invalid": }',
        'null',
    ]
    
    for test in test_cases:
        is_valid = validate_input(test)
        print(f"{test[:20]}... → {'Valid' if is_valid else 'Invalid'}")

if __name__ == "__main__":
    main()